src/formats/bbcode.js

```diff
@@ -2505,15 +2507,10 @@
 						if (isValidChild) {
 							// code tags should skip most styles
 							if (tag !== 'code') {
-								// handle inline bbcodes
 								curTag = handleStyles(node, curTag);
-								curTag = handleTags(node, curTag);
-
-								// handle blocklevel bbcodes
-								curTag = handleStyles(node, curTag, true);
 							}
 
-							curTag = handleTags(node, curTag, true);
+							curTag = handleTags(node, curTag);
```

Comment:
Inlines need to be converted before blocks otherwise it can cause issues with blocks being nested inside of inlines.

I'm not sure if the issue exists with the default BBCodes but for example if you rename the `b` BBCode to `z`:

```js
sceditor.formats.bbcode.set('z', sceditor.formats.bbcode.get('b'));
sceditor.formats.bbcode.remove('b');
```

then create the editor and convert:

```html
<div style="font-weight: bold; text-align: center;">test</div>
```

it will be converted to `[b][center]test[/center][/b]` instead of `[center][b]test[/b][/center]` which it currently does. Thinking about it, there should really be a test to cover this.

Reply:
My goal with redoing this code was to make it more concise and easy to read. The double calls for blocks and inlines annoyed me,. That bug looks to be fixed by separating the styes back out and not the tags https://github.com/samclarke/SCEditor/pull/821/commits/58fab8598aa92ece8db59c9ed01813166dd21dbe

Reply:
I have added tests to ensure proper nesting for both styles and tags.

Reply:
The duplicate calls annoy me too. One thing I wanted to do at some point to simplify the code is move styles in as an attribute (#589) so defining a BBCode would be:

```js
// Match any tag with text-align (equivalent to current styles behaviour)
'*' {
  style: {
    'text-align': null
  }
}
```
or
```js
// Match any p tag with text-align (currently requires matching one or the other and
// manually checking in the format function)
'p' {
  style: {
    'text-align': null
  }
}
```

That way only `handleTags` has to be called twice and `handleStyles` gets remove completely. `handleTags` tags would need to special case the `style` attribute and check for a match in whatever the current matching mode is but even so, that should simplify things. It would also be backwards compatible as we can simply move styles to a * tag as above.

Thinking about it, that might be worth doing for this as currently match all will either matches all styles or all the tags, not both. E.g.:

```js
sceditor.formats.bbcode.set('margin', {
	tags: {
		'p': null
	},
	styles: {
		'margin': null
	},
	isInline: false,
	matchAttrs: sceditor.BBCodeParser.MatchingMode.all,
	format: function (element, content) {
		return '[margin=' + element.style.margin + ']' +
			content + '[/margin]';
	},
	html: '<p style="margin:{defaultattr}">{0}</p>'
});
```

With:

```html
<p>No match</p>
<div style="margin: 1em;">No match</div>
<p style="margin: 1em;">Match</p>
```

Produces:

```
[margin=0px]No match[/margin]
[margin=14px]No match[/margin]
[margin=14px][margin=14px]Match[/margin]
[/margin]
```

If it was done then potentially there could even be multiple matching modes:

`'any'` (current)
`'all-attrs'` (all attributes, any styles)
`'all-styles'` (any attributes, all styles)
`'all'` (all styles and all attributes)

Not sure if that's a good idea as it does look a bit confusing. I definitely think being able to match the style with the tag would be better though.

Reply:
I definitely had #589 in mind when I simplified the data structures. And since `matchAttrs` only affects attributes, your example stays the same regardless of the matching mode.

Matching modes for styles should perhaps be its own property, `matchStyles`, something like that. Matching all tags would probably require some code refactoring as the current setup doesn't look      like it'll facilitate that well. Both features should be separate I think.

Putting styles under tags should be its own PR. It's out of scope of this one.